command/operator_diagnose.go

```diff
@@ -173,119 +186,270 @@ func (c *OperatorDiagnoseCommand) offlineDiagnostics(ctx context.Context) error
+				dirAccess := diagnose.ConsulDirectAccess(config.Storage.Config)
+				if dirAccess != "" {
+					diagnose.Warn(ctx, dirAccess)
+				}
+				return nil
+			})
+		}
+
+		// Attempt to use storage backend
+		if !c.skipEndEnd {
+			diagnose.Test(ctx, "test-access-storage", diagnose.WithTimeout(30*time.Second, func(ctx context.Context) error {
+				// TODO: A static file in storage that probably won't cause a collision seems low-risk to write to for now.
+				// Should we make this a proper uuid?
+
+				veryRandomUuid := "diagnose-secret-uuid-1234"
+				err := diagnose.EndToEndLatencyCheckWrite(ctx, veryRandomUuid, *backend)
+				if err != nil && strings.Contains(err.Error(), diagnose.LatencyWarning) {
+					diagnose.Warn(ctx, err.Error())
+				} else if err != nil {
+					return err
+				}
+				err = diagnose.EndToEndLatencyCheckRead(ctx, veryRandomUuid, *backend)
+				if err != nil && strings.Contains(err.Error(), diagnose.LatencyWarning) {
+					diagnose.Warn(ctx, err.Error())
+				} else if err != nil {
+					return err
+				}
+				err = diagnose.EndToEndLatencyCheckDelete(ctx, veryRandomUuid, *backend)
+				if err != nil && strings.Contains(err.Error(), diagnose.LatencyWarning) {
+					diagnose.Warn(ctx, err.Error())
+				} else if err != nil {
+					return err
+				}
+				return nil
+			}))
+		}
+		return nil
+	})
 
-			// Check ciphersuite and load ca/cert/key files
-			// TODO: TLSConfig returns a reloadFunc and a TLSConfig. We can use this to
-			// perform an active probe.
-			_, _, err := listenerutil.TLSConfig(ln.Config, make(map[string]string), c.UI)
+	var configSR sr.ServiceRegistration
+	diagnose.Test(ctx, "service-discovery", func(ctx context.Context) error {
+		if config.ServiceRegistration == nil || config.ServiceRegistration.Config == nil {
+			return fmt.Errorf("No service registration config")
+		}
+		srConfig := config.ServiceRegistration.Config
+
+		diagnose.Test(ctx, "test-serviceregistration-tls-consul", func(ctx context.Context) error {
+			// SetupSecureTLS for service discovery uses the same cert and key to set up physical
+			// storage. See the consul package in physical for details.
+			err = srconsul.SetupSecureTLS(api.DefaultConfig(), srConfig, server.logger, true)
 			if err != nil {
 				return err
 			}
+			return nil
+		})
 
-			sanitizedListeners = append(sanitizedListeners, listenerutil.Listener{
-				Listener: ln.Listener,
-				Config:   ln.Config,
+		if config.ServiceRegistration != nil && config.ServiceRegistration.Type == "consul" {
+			diagnose.Test(ctx, "test-consul-direct-access-service-discovery", func(ctx context.Context) error {
+				dirAccess := diagnose.ConsulDirectAccess(config.ServiceRegistration.Config)
+				if dirAccess != "" {
+					diagnose.Warn(ctx, dirAccess)
+				}
+				return nil
 			})
 		}
-		return diagnose.ListenerChecks(sanitizedListeners)
-	}); err != nil {
-		return err
+		return nil
+	})
+
+	sealcontext, sealspan := diagnose.StartSpan(ctx, "create-seal")
```

Comment:
This is just the modularization part right?  Testing the seal encrypt/decrypt is later?

Reply:
The seal encrypt/decrypt is in a separate PR (https://github.com/hashicorp/vault/pull/11627). This creates all the seals specified in the config file.